orchagent/portsorch.cpp

```diff
@@ -6095,58 +6094,134 @@ void PortsOrch::generateQueueMapPerPort(const Port& port, FlexCounterQueueStates
         {
             queuePortVector.emplace_back(id, sai_serialize_object_id(port.m_port_id));
         }
+    }
+
+    if (voq)
+    {
+        m_voqTable->set("", queueVector);
+    }
+    else
+    {
+        m_queueTable->set("", queueVector);
+    }
+    m_queuePortTable->set("", queuePortVector);
+    m_queueIndexTable->set("", queueIndexVector);
+    m_queueTypeTable->set("", queueTypeVector);
+
+    CounterCheckOrch::getInstance().addPort(port);
+}
+
+void PortsOrch::addQueueFlexCounters(map<string, FlexCounterQueueStates> queuesStateVector)
+{
+    if (m_isQueueFlexCountersAdded)
+    {
+        return;
+    }
 
-        // Install a flex counter for this queue to track stats
-        std::unordered_set<string> counter_stats;
-        for (const auto& it: queue_stat_ids)
+    for (const auto& it: m_portList)
+    {
+        if (it.second.m_type == Port::PHY)
         {
-            counter_stats.emplace(sai_serialize_queue_stat(it));
+            addQueueFlexCountersPerPort(it.second, queuesStateVector.at(it.second.m_alias));
         }
-        queue_stat_manager.setCounterIdList(queue_ids[queueIndex], CounterType::QUEUE, counter_stats);
+    }
 
-        if (voq)
+    m_isQueueFlexCountersAdded = true;
+}
+
+
+void PortsOrch::addQueueFlexCountersPerPort(const Port& port, FlexCounterQueueStates& queuesState)
+{
+    for (size_t queueIndex = 0; queueIndex < port.m_queue_ids.size(); ++queueIndex)
+    {
+        string queueType;
+        uint8_t queueRealIndex = 0;
+        if (getQueueTypeAndIndex(port.m_queue_ids[queueIndex], queueType, queueRealIndex))
```

Comment:
can we check `(!queuesState.isQueueCounterEnabled(queueRealIndex))` first and then `(getQueueTypeAndIndex(port.m_queue_ids[queueIndex], queueType, queueRealIndex))`?
AFAIU, the former is light and the latter is a bit heavier. If the former returns false, we do not need to call the latter.

Reply:
how can this be done ?
we first get the queueRealIIndex using the getQueueTypeAndIndex and then use in for queuesState.isQueueCounterEnabled(queueRealIndex).
this is also the implementation of the PR #2473 [counters] Improve performance by polling only configured ports buffer queue/pg counters that uses it in generateQueueMapPerPort function.
can I use the queueIndex rather than the queueRealIndex ? if so I can ditch the whole getQueueTypeAndIndex if statement.

Reply:
Ok. You are correct. I didn’t realize it depends on the output argument of the former function call.